### PR TITLE
STACK : 526 - A10-Octavia Configuration file

### PR DESCRIPTION
**Pull request by afrin-chakure**
## Customer Highlights 

Now a10-octavia will have its own configuration option support for following sections 

-  `a10_controller_worker`
- `a10_house_keeping`
- `a10_health_manager`
- `a10_nova`

- `vthunder`
- `slb`
- `listener`
- `service_group`
- `l7policy`
- `l7rule`
- `server`
- `rack_vthunder`

## Non-Customer Highlights

a) Following new files are added, for customizing configuration for above sections in `a10-octavia.conf` file that should be located at `/etc/a10/`

files : 
`a10_octavia/common/config_options.py`
`a10_octavia/common/config_types.py`
`a10_octavia/common/exceptions.py`
`a10_octavia/common/utils.py`

b) The `config_types` module adds a support for type - **list of objects(dict)**  in configurations  
eg - 
```
devices=[{ "field1" : "value1", 
                  "field2" : "value2", 
                  "field3" : "value3"},
                { "field1" : "value1", 
                  "field2" : "value2", 
                  "field3" : "value3" }]
```
Such configuration exist for [rack_vthunder] section. Also, this module will have validation methods related to `rack_vthunder.devices` params.

c) The existing `a10_octavia/a10_config.py` module had dependencies that is no longer required after adding this new way of accessing configuration. Removed them.

d)  `a10_octavia/cmd/service.py`  module will call our new  `config_options` module for initializing global configuration.

e) `a10_octavia/install/a10-*.service`, defines the location of our configuration directory and files.
```--config-dir=/etc/octavia --config-file=/etc/a10/a10-octavia.conf```
All `*.conf` files will be picked up from mentioned directories. Hence we will have octavia's default configuration as well.

f) pep8 and pylint validations done for new files. 
image

g) Added docstrings for new files
h) Updated README accordingly for a10-octavia project.

## References 
a) To add VThunder related sections, help documentation reference for parameters is taken from `/aab/docs/` 
b) For other sections, parameter references are taken from existing `/octavia/octavia/common/config.py` module.
c) For new config support -> `https://github.com/openstack/oslo.config` - `cfg.py` and `types.py` modules.

## Testing
Tested with following sample config file :
[a10-octavia-conf.txt](https://github.com/a10networks/a10-octavia/files/4110230/a10-octavia-conf.txt)

Some testing screenshots : 
a) template configuration for slb nouns
slb-noun

vthunder configured

image

b) Successful rack vthunder configuration via conf
rack_config

c) Error for wrong rack configs in file
error_wrong_conf

for invalid configuration
image

d) Warning for invalid ip_address in [rack_vthunder] section
warnings-wrong_ip_address

e) Our `a10-octavia.conf` should reside at `/etc/a10/` location, else `ConfigFilesNotFoundError` is thrown during restart of service.
image

## Limitations/Notes
a) Configurations like `enabled_provider_drivers` and `default_provider_driver` under `api_settings`, needs to given in the main `octavia.conf` file itself, otherwise it won't pick up the required `a10` driver.
b) Pylint and pep8 only done for new files and not for existing files. It will be taken care during refactoring stories.

## Closes 
[STACK-526](https://a10networks.atlassian.net/browse/STACK-526)
SubTasks - [STACK-557](https://a10networks.atlassian.net/browse/STACK-557) ,  [STACK-558](https://a10networks.atlassian.net/browse/STACK-558) , [STACK-559](https://a10networks.atlassian.net/browse/STACK-559)

### CODE REVIEW

File: a10_octavia/common/a10_config.py [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+db_options.set_defaults(cfg.CONF, connection=_SQL_CONNECTION_DEFAULT,
+                        max_pool_size=10, max_overflow=20, pool_timeout=10)
+
+logging.register_options(cfg.CONF)
+
+ks_loading.register_auth_conf_options(cfg.CONF, constants.SERVICE_AUTH)
+ks_loading.register_session_conf_options(cfg.CONF, constants.SERVICE_AUTH)
+
+
+def init(args, **kwargs):
```
Comment (hthompson6) [inline a10_octavia/common/a10_config.py]: This should be `*args`

File: a10_octavia/common/a10_types.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            if validation_flag:
+                rack_device['undercloud'] = True
+                vthunder_conf = data_models.VThunder(**rack_device)
+                rack_dict[rack_device['project_id']] = vthunder_conf
+            else:
+                LOG.warning('Invalid definition of rack device for '
+                            'project ' + rack_device['project_id'])
+
+    except KeyError as err:
+        LOG.error("Invalid definition of rack device in configuration file."
```
Comment (hthompson6) [inline a10_octavia/common/a10_types.py]: If they misconfigure the rack device then we should not boot anything.

File: a10_octavia/controller/worker/controller_worker.py [hunk 3]
```diff
@@ -309,17 +306,17 @@ def create_load_balancer(self, load_balancer_id):
                  constants.BUILD_TYPE_PRIORITY:
                  constants.LB_CREATE_NORMAL_PRIORITY}
 
-        topology = CONF.controller_worker.loadbalancer_topology
-
+        topology = CONF.a10_controller_worker.loadbalancer_topology
+
```
Comment (hthompson6) [inline a10_octavia/controller/worker/controller_worker.py]: White space

File: a10_octavia/controller/worker/controller_worker.py [hunk 4]
```diff
@@ -208,7 +205,7 @@ def create_listener(self, listener_id):
 
         load_balancer = listener.load_balancer
 
-        if listener.project_id in self.rack_dict:
+        if listener.project_id in CONF.RACK_VTHUNDER.devices:
```
Comment (hthompson6) [inline a10_octavia/controller/worker/controller_worker.py]: `RACK_VTHUNDER` capitalizing this is improper. We should keep consistent and use lower case with underscore separators.

File: a10_octavia/controller/worker/tasks/a10_database_tasks.py [hunk 5]
```diff
@@ -67,14 +65,11 @@ class CreteVthunderEntry(BaseDatabaseTask):
...
         vthunder_id = uuidutils.generate_uuid()
-        self.config = self.a10_conf.get_conf()
 
-        username = self.config.get(
-            'DEFAULT', 'DEFAULT_VTHUNDER_USERNAME').replace('"', '')
-        password = self.config.get(
-            'DEFAULT', 'DEFAULT_VTHUNDER_PASSWORD').replace('"', '')
-        axapi_version = int(
-            self.config.get('DEFAULT', 'DEFAULT_AXAPI_VERSION').replace('"', ''))
+        username = CONF.VTHUNDER.DEFAULT_VTHUNDER_USERNAME
```
Comment (hthompson6) [inline a10_octavia/controller/worker/tasks/a10_database_tasks.py]: Only `conf` should be capitalized. As per how it's done everywhere else in the Octavia codebase.

File: a10_octavia/controller/worker/tasks/handler_server.py [hunk 6]
```diff
@@ -28,39 +28,22 @@ class MemberCreate(BaseVThunderTask):
     def execute(self, member, vthunder, pool):
         """Execute create member for an amphora."""
 
-        conn_limit = self.readConf('SERVER', 'conn_limit')
-        if conn_limit is not None:
-            conn_limit = int(conn_limit)
-        conn_resume = self.readConf('SERVER', 'conn_resume')
-        if conn_resume is not None:
-            conn_resume = int(conn_resume)
+        conn_limit = CONF.SERVER.conn_limit
```
Comment (hthompson6) [inline a10_octavia/controller/worker/tasks/handler_server.py]: `server` shouldn't be capitalized....

File: a10_octavia/common/a10_types.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    :param item_type: type of item (None)
+    :param bounds: if True the value should be inside "[" and "]" pair
+    :param \\*\\*kwargs: arbitrary keyword arguments passed to :class:`Opt`
+
+    """
+    def __init__(self, name, item_type=None, bounds=None, **kwargs):
+        super(ListOfDictOpt, self).__init__(name, type=ListOfObjects(item_type, bounds), **kwargs)
+
+
+class ListOfObjects(List):
```
Comment (hthompson6) [inline a10_octavia/common/a10_types.py]: Good and proper use of inheritance here 👍

File: a10_octavia/common/a10_types.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+        final_list = []
+        for item in value:
+            item = '{' + item.lstrip('{').rstrip('}') + '}'
+            item = item.replace("\'", "\"")
+            single_dict = json.loads(item)
+            final_list.append(single_dict)
+        return convert_to_rack_vthunder_conf(final_list)
+
+
+def convert_to_rack_vthunder_conf(rack_list):
```
Comment (hthompson6) [inline a10_octavia/common/a10_types.py]: In most cases, it's better to define outside functions above where they are being used for readability sake.

As you can see here: https://github.com/openstack/python-neutronclient/blob/master/neutronclient/v2_0/client.py#L47

Comment (hthompson6) [reply a10_octavia/common/a10_types.py]: Also naming things before they are used removes change of edge case errors such as the following:
https://stackoverflow.com/questions/18794873/should-function-be-defined-before-it-is-used-in-python

Comment (hthompson6) [reply a10_octavia/common/a10_types.py]: If you find that you have lots of helper functions being used in multiple places which is making it hard to find and read what the classes are doing then it's time to create a `utils.py` or `helper.py` function that can be leveraged instead.

**Comment by hthompson6** [comment]
VTHUNDER, LISTENER, SLB, L7POLICY, L7RULE all need to be made lower case.

File: a10_octavia/common/a10_types.py (hunk 2 above)
Comment (afrin-chakure) [reply a10_octavia/common/a10_types.py]: Had discussion with Omkar on this. If they misconfigure rack conf, it should boot up a vthunder.

File: a10_octavia/common/a10_types.py (hunk 8 above)
Comment (afrin-chakure) [reply a10_octavia/common/a10_types.py]: Shifted all the utils methods to a common module.

**Pull request merged**